python/ray/ml/train/data_parallel_trainer.py

```diff
@@ -0,0 +1,184 @@
+import logging
+from typing import Dict, Callable, Optional, Union
+
+from ray.ml.trainer import Trainer
+from ray.ml.config import ScalingConfig, RunConfig
+from ray.ml.train.trainer import GenDataset
+from ray.ml.preprocessor import Preprocessor
+from ray.ml.checkpoint import Checkpoint
+from ray.train import BackendConfig
+from ray.util.annotations import DeveloperAPI
+
+logger = logging.getLogger(__name__)
+
+
+@DeveloperAPI
+class DataParallelTrainer(Trainer):
+    """A Trainer for data parallel training.
+
+    You should subclass this Trainer if your Trainer follows SPMD (single program,
+    multiple data) programming paradigm- you want multiple processes to run the same
+    function, but on different data.
+
+    This Trainer runs the function ``train_loop_per_worker`` on multiple Ray
+    Actors. The provided datasets are automatically sharded to each Ray actor running
+    the training function.
+
+    The ``train_loop_per_worker`` function is expected to take in either 0 or 1
+    arguments:
+
+    .. code-block:: python
+
+        def train_loop_per_worker():
+            ...
+
+    .. code-block:: python
+
+        def train_loop_per_worker(config: Dict):
+            ...
+
+    If ``train_loop_per_worker`` accepts an argument, then
+    ``train_loop_config`` will be passed in as the argument. This is useful if you
+    want to tune the values in ``train_loop_config`` as hyperparameters.
+
+    If the ``datasets`` dict contains a training dataset (denoted by
+    the "train" key), then it will be split into multiple dataset
+    shards that can be consumed by ``train_loop_per_worker``. All the other datasets
+    will not be sharded.
+
+    Inside the ``train_loop_per_worker`` function, you can use any of the
+    :ref:`Ray Train function utils <train-api-func-utils>`.
+
+    .. code-block:: python
+
+        def train_loop_per_worker():
+            # Report intermediate results for callbacks or logging.
+            train.report(...)
+
+            # Checkpoints the provided dict as restorable state.
+            train.save_checkpoint(...)
+
+            # Returns dict of last saved checkpoint.
+            train.load_checkpoint()
+
+            # Returns the Ray Dataset shard for the given key.
+            train.get_dataset_shard("my_dataset")
+
+            # Total number of workers executing training.
+            train.get_world_size()
+
+            # Rank of this worker.
+            train.get_world_rank()
+
+            # Rank of the worker on the current node.
+            train.get_local_rank()
+
+    How do I use ``DataParallelTrainer`` or any of its subclasses?
+
+        Example:
+            .. code-block:: python
+
+                import ray
+                from ray import train
+
+                def train_loop_for_worker():
+                    dataset_shard_for_this_worker = train.get_dataset_shard("train")
+
+                    assert len(dataset_shard_for_this_worker) == 1
```

Comment:
Can be done in a follow-up PR, but I don't think this example sufficiently answers the question. 😅

Reply:
Can you clarify more here?